### PR TITLE
Improve EnemyBallPlacementPlay and convert it to use FSMs

### LINKED ISSUE #2599: Change Enemy Ball Placement Play to shadow less aggressively
### Description of the task
Currently the Enemy Ball Placement Play can be quite aggressive when shadowing the enemy robot placing the ball. This leads to the placing robot being impeded, which can result in a penalty according to the SSL rulebook. It seems that shadowing is most aggressive when the enemy robot is moving towards the ball to pick it up. After the enemy robot picks up the ball, we no longer shadow as aggressively.

See second test in `ball_placement_play_test.py` for more details

### Acceptance criteria

- [ ] Enemy Ball Placement Play does not impede robot which is placing ball
- [ ] Parameterized test case in `ball_placement_play_test.py` no longer has `test_duration` as a paramter - All placement should be done within a reasonable amount of time (10 - 15 seconds)

### Blocked By
Sort of blocked by #2558. My first PR on #2558 should unblock this.

### LINKED ISSUE #2783: Validation for Enemy Ball Placement play
### Description of the task
Currently, `enemy_ball_placement_play_test.py` is lacking validation. Validation functions may need to be written, and definitely need to be added to the test cases in the file. A good start would be too look at `enemy_ball_placement_play.cpp`

### Acceptance criteria

- [ ] Always validation for EBP
- [ ] Eventually validation for EBP

### LINKED ISSUE #2870: Fix AI behaviour on enemy_ball_placement
### Description of the task
The goal of this ticket is to ensure that our AI satisfies the requirements of robocup's ball placement rules listed https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement_interference. 

Currently, our design did not follow the specifications and exhibit weird behavior. A validation has already been created in #2783  and that could be used to verify the design. 

### Acceptance criteria

- [ ] pass all enemy_ball_placement_validation
- [ ] follows rules documented in ssl rules: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement_interference

### LINKED ISSUE #3018: [Gameplay] Improve EnemyBallPlacementPlay and convert it to using FSMs
### Description of the task

The [enemy ball placement rules](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement_interference) state that robots should always maintain a distance of at least 0.5 meters from the line between the ball and the ball placement point when the enemy team is performing ball placement. 

At Robocup 2023, our EnemyBallPlacementPlay attempted to shadow the enemy robot as it performed ball placement. On multiple occasions, our robots violated the enemy ball placement rule, which incurred several fouls. 

We should rewrite EnemyBallPlacementPlay so that our robots do not interfere with the enemy's robot performing ball placement. 

We should also use FSMs instead of coroutines to implement EnemyBallPlacementPlay. 

### Acceptance criteria

- [ ] Convert EnemyBallPlacementPlay to use FSMs instead of coroutines
- [ ] Ensure that our enemy ball placement behaviour does not violate SSL rules
- [ ] Improve the enemy ball placement tests

### PR DESCRIPTION
**Pull request by raymond212**
### Description

1. Fixed a bug in sensor fusion where the ball placement position is not sent to non-placing team, resulting in the stadium obstacle in #3121 not working. 

2. Rewrote `EnemyBallPlacementPlay` to use the FSM architecture. 

3. Re-designed the play as: if the ball is far from the placement point, then focus on avoiding ball placement interference by assigning AvoidInterferenceTactics (clones of MoveTactic) to move robots out of the way. AvoidInterferenceTactics override motion constraints, because sometimes the stadium motion constraint can actually trap robots and prevent them from moving to somewhere that doesn't interfere. 

   Once the ball is nearly placed, then we switch to a defensive formation with 1 goalie, 2 crease defenders, and 3 robots to stay near the ball without interfering (essentially our StopPlay behaviour). By doing this, we can transition into our other plays quickly once the game has been resumed.  

Placement1

Placement2

The edge case where the stadium goes through the defense area is also handled well. 

PlacementThroughDefenseArea1

In the first and third gifs, there is a sensor fusion bug where the ball position does not get updated, possibly because the yellow placement robot is blocking the ball. This could be something to address in a separate issue. 

4. Added validation to check that we do not have robots stay in the ball placement stadium for over two seconds. 

~~Works well generally. Three robots stay near the ball without interfering, so that we can quickly transition into the play that follows, while we also have defensive robots as the goalie and crease defenders.~~

~~However, there are issues with some edge cases. Currently, I try to position the three robots on the other side of a line connecting the ball to the ball placement position. This could end up outside the field, which is problematic. To fix this, I think I will try to directly send these robots near the final ball placement point, while making sure to stay out of the stadium obstacle.~~

~~Also, the stadium obstacle is not fully effective at keeping out the goalie and crease defenders, in the edge case that the stadium obstacle goes right through our defense zone. While unlikely, this is technically possible and allowed under robocup rules [5.2](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement). The best solution for this situation would probably be to use move tactics to force our defensive robots out of the stadium.~~

### Testing Done

Tested in Thunderscope and `enemy_ball_placement_play_test.py`. 

### Resolved Issues

resolves #3018, resolves #2783, resolves #2870, resolves #2599

### Length Justification and Key Files to Review

* `enemy_ball_placement_play_fsm.h` and `enemy_ball_placement_play_fsm.cpp` for FSM implementation
* `robot_enters_placement_region.py` for validation

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by nimazareian** [comment]
Good work on finding this bug!

With regards to this
> Also, the stadium obstacle is not fully effective at keeping out the goalie and crease defenders, in the edge case that the stadium obstacle goes right through our defense zone. While unlikely, this is technically possible and allowed under robocup rules [5.2](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement). The best solution for this situation would probably be to use move tactics to force our defensive robots out of the stadium.

I have a work in progress branch with a couple of core changes to the trajectory planner, including integrating #3058 into the planner. This _should_ ™️ move the robots out of all non-robot obstacles, so hopefully all robots will move out of the stadium. You may have already done this, but it might be helpful to have a sim test with this edge case that I could run on my branch later as well. 

If the ball moves quick, causing the stadium to move quickly, we might still receive a foul... but Im not sure if that is really avoidable.

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    move_tactics[0]->updateControlParams(
+        center, positioning_vector.orientation() + Angle::half(), 0);
+    move_tactics[1]->updateControlParams(left, left_vector.orientation() + Angle::half(),
+                                         0);
+    move_tactics[2]->updateControlParams(right,
+                                         right_vector.orientation() + Angle::half(), 0);
+    tactics_to_run[0].emplace_back(move_tactics[0]);
+    tactics_to_run[0].emplace_back(move_tactics[1]);
+    tactics_to_run[0].emplace_back(move_tactics[2]);
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: This would be interesting to see. We just gotta becareful of not really following the enemy robot placing the ball, as it could make it very easy to accidently get a foul if we get too close.

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      crease_defenders({std::make_shared<CreaseDefenderTactic>(
+                            ai_config.robot_navigation_obstacle_config()),
+                        std::make_shared<CreaseDefenderTactic>(
+                            ai_config.robot_navigation_obstacle_config())}),
+      move_tactics({
+          std::make_shared<MoveTactic>(),
+          std::make_shared<MoveTactic>(),
+          std::make_shared<MoveTactic>(),
+      }),
+      distance_to_keep(ENEMY_BALL_PLACEMENT_DISTANCE_METERS + 0.2)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: To be safe, it might be a good idea to increase this a bit further by also taking into account the robot radius

**Review by nimazareian — COMMENTED** [review]
I know this PR is still a draft, but quickly looked over it to give some initial feedback. Overall it looks great! 

Also, if you haven't already, I would recommend testing this out with autoref as well! #2788

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_test.py
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_test.py:1]: please add validation to this file

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                                         0);
+    move_tactics[2]->updateControlParams(right,
+                                         right_vector.orientation() + Angle::half(), 0);
+    tactics_to_run[0].emplace_back(move_tactics[0]);
+    tactics_to_run[0].emplace_back(move_tactics[1]);
+    tactics_to_run[0].emplace_back(move_tactics[2]);
+
+    // Set tactics to run
+    event.common.set_tactics(tactics_to_run);
+}
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: how about an alternate implementation?

- we assign 6 move tactics
- for each tactic, the robot stays in place unless it is inside the stadium obstacle
- if it intersects the obstacle, we move to the closest point toward the centre of the field

20240401_082244515_iOS

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_test.py
Comment (itsarune) [reply src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_test.py:1]: you should also add more tests

**Review by itsarune — COMMENTED** [review]
nice work, looking good... left some comments about an alternate implementation that addresses your edge case that you found

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp (hunk 3 above)
Comment (raymond212) [reply src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: That seems like an effective way of handling the edge case. However, I pulled from upstream and after some testing, I think Nima's #3152 PR fixes the issue of robots driving into the stadium obstacle, so the edge case of the goalie and crease defenders being in the stadium isn't really an issue anymore. I will work on adding tests for now, does that sound good?

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    tactics_to_run[0].emplace_back(crease_defenders[0]);
+    tactics_to_run[0].emplace_back(crease_defenders[1]);
+
+    // Create move tactics
+    Vector positioning_vector = (world_ptr->ball().position() - placement_point);
+
+    // If the ball is nearly placed, then adjust move tactics to position between
+    // friendly goal and the ball
+    if (positioning_vector.length() < 0.5)
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: 0.5 should be a static constant

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        for robot in world.friendly_team.team_robots:
+            if tbots_cpp.contains(
+                stadium, tbots_cpp.createPoint(robot.current_state.global_position)
+            ):
+                if not self.first_enter_time:
+                    self.first_enter_time = time.time()
+                else:
+                    # By SSL rules, only staying in the stadium for over 2 seconds counts
+                    if time.time() - self.first_enter_time > 2:
```
Comment (itsarune) [inline src/software/simulated_tests/robot_enters_placement_region.py]: the `2` should be initialized in the constructor

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                        return ValidationStatus.PASSING
+                break
+        else:
+            self.first_enter_time = None
```
Comment (itsarune) [inline src/software/simulated_tests/robot_enters_placement_region.py]: this `else` is out-of-line with the rest of the function

Comment (itsarune) [reply src/software/simulated_tests/robot_enters_placement_region.py]: this logic doesn't quite make sense. `first_enter_time` should only be set to `None` if none of the robots are in the ball placement region and it doesn't look like that's happening here.

Comment (raymond212) [reply src/software/simulated_tests/robot_enters_placement_region.py]: The out of line `else` will run if the loop executes normally without `break`. 

The only way for the loop to execute without breaking is for there to be no robots in the stadium, which is why `first_enter_time` gets set to `None`.

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp (hunk 3 above)
Comment (raymond212) [reply src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: Incorporated into the new implementation.

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.h [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+#include "software/geom/stadium.h"
+#include "software/world/game_state.h"
+
+/**
+ * This FSM implements the enemy ball placement play.
+ * - If the ball placement point does not exist yet, then we wait
+ * - If the ball placement point exists, but the ball is far from the placement point,
+ * then we focus on avoiding ball placement interference If our robot R is in the ball
+ * placement stadium, then it will move perpendicular to the segment connecting * and +,
+ * to either p1 or p2, whichever one is within the field boundaries and closest to R
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.h]: nit
```suggestion
 * - If the ball placement point exists, but the ball is far from the placement point,
 * then we focus on avoiding ball placement interference. If our robot R is in the ball
 * placement stadium, then it will move perpendicular to the segment connecting * and +,
 * to either p1 or p2, whichever one is within the field boundaries and closest to R
```

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      }),
+      move_tactics({
+          std::make_shared<MoveTactic>(),
+          std::make_shared<MoveTactic>(),
+          std::make_shared<MoveTactic>(),
+      }),
+      goalie_tactic(std::make_shared<GoalieTactic>(ai_config)),
+      distance_to_keep(ENEMY_BALL_PLACEMENT_DISTANCE_METERS +
+                       4 * ROBOT_MAX_RADIUS_METERS),
+      nearly_placed_threshold(0.5)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: nit: Instead of having fields defined in this class, could you add the constants to `parameters.proto` (could be a new message for enemy ball placement) so they are easily configurable during a game? The new parameters should also have the unit of distance in their name as well.

For `distance_to_keep` you could probably make the dynamic parameter represent the additional distance to keep. i.e. 0m represents 0.5m away and 0.2m represents represents 0.7m away, ...

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            }
+            avoid_interference_tactics[idx]->updateControlParams(
+                destination, (ball_pos - robot.position()).orientation(), 0);
+        }
+        else
+        {
+            avoid_interference_tactics[idx]->updateControlParams(
+                robot.position(), (ball_pos - robot.position()).orientation(), 0);
+        }
+        tactics_to_run[0].emplace_back(avoid_interference_tactics[idx++]);
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: nit: For easier readability, I think it might be better to increment `idx` in a separate line as opposed to inline

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            else if (!contains(fieldLines, p2))
+            {
+                destination = p1;
+            }
+            else
+            {
+                destination =
+                    distance(p1, robot.position()) < distance(p2, robot.position()) ? p1
+                                                                                    : p2;
+            }
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: nit: we've generally been staying away from using ternary operators in C++ for readability. you could split this up to by adding another else if to your existing checks.

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            avoid_interference_tactics[idx]->updateControlParams(
+                destination, (ball_pos - robot.position()).orientation(), 0);
+        }
+        else
+        {
+            avoid_interference_tactics[idx]->updateControlParams(
+                robot.position(), (ball_pos - robot.position()).orientation(), 0);
+        }
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: Could you add some comments clarifying what youre calculating and why youre deciding on which point to pick as the destination?

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class RobotEntersPlacementRegion(Validation):
+
+    """Checks if a Robot enters the ball placement stadium region."""
+
+    def __init__(self, placement_point):
+        self.placement_point = placement_point
+        self.first_enter_time = None
+        self.interference_radius = 0.5
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py]: nit: Could you update this to use `ENEMY_BALL_PLACEMENT_DISTANCE_METERS` instead (you will need to add it to `py_constants.cpp` as well)?

Likewise in this existing test as well:
https://github.com/UBC-Thunderbots/Software/blob/426e7bc93a98d6e31ab306da31916c4fdfc8894d/src/software/ai/hl/stp/play/ball_placement/ball_placement_play_test.py#L129-L138

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        self.stay_in_region_threshold = (
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py]: nit:
```suggestion
        self.stay_in_region_threshold_sec = (
```

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            2  # By SSL rules, only staying in the stadium for over 2 seconds counts
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py]: Could you also add a link to the rule: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement_interference

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    create_validation_geometry,
+    create_validation_types,
+)
+
+
+class RobotEntersPlacementRegion(Validation):
+
+    """Checks if a Robot enters the ball placement stadium region."""
+
+    def __init__(self, placement_point):
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py:16]: nit: missing function docs

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+                stadium, tbots_cpp.createPoint(robot.current_state.global_position)
+            ):
+                if not self.first_enter_time:
+                    self.first_enter_time = time.time()
+                else:
+                    if (
+                        time.time() - self.first_enter_time
+                        > self.stay_in_region_threshold
+                    ):
+                        return ValidationStatus.PASSING
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py]: We should generally avoid using actual wall time in our simulation to allow for the test to run faster or slower than realtime reliably. Though I'm not sure if there currently exists a better way of getting the current simulation times. Tagging @itsarune @sauravbanna @williamckha in case they know of a better way. If not, then we should track this as an issue to be fixed in a future PR.

File: src/software/simulated_tests/robot_enters_placement_region.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+    def get_validation_geometry(self, world) -> ValidationGeometry:
+        """
+        (override) shows regions to enter
+        """
+        segment = tbots_cpp.Segment(
+            self.placement_point,
+            tbots_cpp.createPoint(world.ball.current_state.global_position),
+        )
+        stadium = tbots_cpp.Stadium(segment, 0.5)
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_placement_region.py]: nit: 
```suggestion
        stadium = tbots_cpp.Stadium(segment, ENEMY_BALL_PLACEMENT_DISTANCE_METERS)
```

**Review by nimazareian — COMMENTED** [review]
Just tried the new implementation with auto-ref and it looks great! I'm excited to get this in and use autoref more often now. :yellow_square: 

I just had a few nits.

**Review by raymond212 — COMMENTED** [review]
Addressed Nima's comments, new changes are ready for review.

File: src/proto/parameters.proto [hunk 18]
```diff
@@ -37,6 +37,7 @@ message AiConfig
     required EnemyCapabilityConfig enemy_capability_config                  = 7;
     required ShootOrPassPlayConfig shoot_or_pass_play_config                = 8;
     required RobotCapabilitiesConfig robot_capabilities_config              = 9;
+    required EnemyBallPlacementPlayConfig enemy_ball_placement_play_config  = 10;
```
Comment (nimazareian) [inline src/proto/parameters.proto:40]: it's interesting that we had skipped field 10

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp [hunk 19]
```diff
@@ -57,39 +57,56 @@ void EnemyBallPlacementPlayFSM::avoid(const Update& event)
...
-                       lateral_positioning_vector * distance_to_keep;
+                       lateral_positioning_unit_vector * distance_to_keep_meters;
             Point p2 = placement_point + longitudinal_positioning_vector -
-                       lateral_positioning_vector * distance_to_keep;
+                       lateral_positioning_unit_vector * distance_to_keep_meters;
             Point destination;
             Rectangle fieldLines = world_ptr->field().fieldLines();
-            if (!contains(fieldLines, p1))
+            // If either destination point is outside the field, then pick the other
+            // point. It is impossible for both points to be outside the field
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.cpp]: Isn't it technically possible for both p1 and p2 (shown in red) to be outside of the field if the ball is in the corner and the placement position is somewhere towards the center? (the straight line shows the "backbone")
image

Though in either case we can _probably_ ignore this edge case

**Review by nimazareian — DISMISSED** [review]
LGTM! :soccer: 

Just had two additional non-blocking comments

**Review by nimazareian — DISMISSED** [review]

File: src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.h [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+ *            +
+ *      R   x  enemy robot with ball
+ *        R
+ *          R
+ *
+ *      D   D
+ *
+ *        G
+ *     +-----+
+ */
```
Comment (williamckha) [inline src/software/ai/hl/stp/play/enemy_ball_placement/enemy_ball_placement_play_fsm.h:43]: Helpful diagrams!

**Review by williamckha — APPROVED** [review]
LGTM! Fouls begone

File: src/software/simulated_tests/robot_enters_placement_region.py (hunk 16 above)
Comment (itsarune) [reply src/software/simulated_tests/robot_enters_placement_region.py]: As discussed before, we'd use the timestamp stored in the Wotld proto in case someone else sees this thread.

**Review by nimazareian — APPROVED** [review]
